File: 41. First Missing Positive/41.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        for (int i=0; i<nums.size(); ++i) nums[i] = (nums[i]<0) ? 0 : nums[i];
        for (int i=0; i<nums.size(); ++i)
            if ((nums[i]&INT_MAX)>=1 && (nums[i]&INT_MAX)<=nums.size())
                nums[(nums[i]&INT_MAX)-1] |= INT_MIN;
        for (int i=0; i<nums.size(); ++i) if (!(nums[i]>>31&1)) return i+1;
        return nums.size()+1;
    }
};

File: 0041. First Missing Positive/41.py
class Solution(object):

    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        mask = 1 << 31
        nums = list(map(lambda x: 0 if x < 0 else x, nums))
        for i in nums:
            if (i & ~mask) >= 1 and i & ~mask <= len(nums):
                nums[(i & ~mask) - 1] |= mask
        for i in nums:
            if not(i >> 31 & 1):
                return nums.index(i) + 1
        return len(nums) + 1
